{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The original matrix search was written by H Nakano 20220908.\n",
    "# The search algorithm levaraging base20 indexing was written by Soki Nakano.\n",
    "# Performance improvements, code cleanup and flexible use of any peptide size written by Maurizio Camagna 20230413\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys, csv, re\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Changes made from previous version\n",
    "- The length of the the peptide is now inferred from the number of columns in the *score_table* file\n",
    "- Improved preformance: The previous version calculated all scores for all possible peptides of a given size and stored the scores in a large list. This was inefficent, since scores are calculated for peptides, that don't even occur in the proteom, and with larger peptides, this approach gets exponentially more inefficient. Instead, the scores are now only calculated for the peptides found in the fasta file. Once a score was calculated, it is stored in the list to skip the score calculation the next time the same peptide is encountered\n",
    "- The file *score_ref_name* fulfilled no purpose and was removed (it was only written, but never read)\n",
    "- Unused lines of code were removed\n",
    "- The position is correctly reported in the output file\n",
    "- Added an additional, compact output file\n",
    "    - One row per protein\n",
    "    - Contains total peptide score of each protein\n",
    "    - Peptide matches are written as upper-case in the protein sequence\n",
    "- Added some info on the scoring table (min/max scores)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notes:\n",
    "- The column names of the score table are arbitrary, except for the first column, which must be named **AA**\n",
    "- The order of the rows in the score table is also arbitrary, as the script will sort the file alphabetically"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#score_table_file = 'Input_file_6mer.csv' #name of score table file.\n",
    "score_table_file = 'Input_file.csv' #name of score table file.\n",
    "fasta_in = 'Human_proteins.fasta' # name of  multifasta file for search.\n",
    "minimum_score = 12 # specify minimum score \n",
    "csv_name = 'output.csv' # output file name of Q and its surrounding without Q close to the C-terminal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br><br><br><br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA</th>\n",
       "      <th>-1</th>\n",
       "      <th>Q</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>D</td>\n",
       "      <td>0.627703</td>\n",
       "      <td>0</td>\n",
       "      <td>0.779207</td>\n",
       "      <td>0.795582</td>\n",
       "      <td>0.685065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>0.684350</td>\n",
       "      <td>0</td>\n",
       "      <td>0.598563</td>\n",
       "      <td>0.842955</td>\n",
       "      <td>0.543147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N</td>\n",
       "      <td>0.741804</td>\n",
       "      <td>0</td>\n",
       "      <td>0.701087</td>\n",
       "      <td>0.843187</td>\n",
       "      <td>0.796826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q</td>\n",
       "      <td>1.497140</td>\n",
       "      <td>10</td>\n",
       "      <td>1.035081</td>\n",
       "      <td>1.012754</td>\n",
       "      <td>0.704695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P</td>\n",
       "      <td>0.918332</td>\n",
       "      <td>0</td>\n",
       "      <td>0.567865</td>\n",
       "      <td>0.603739</td>\n",
       "      <td>0.676449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Y</td>\n",
       "      <td>0.964919</td>\n",
       "      <td>0</td>\n",
       "      <td>1.106978</td>\n",
       "      <td>1.679671</td>\n",
       "      <td>0.871349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>W</td>\n",
       "      <td>1.093472</td>\n",
       "      <td>0</td>\n",
       "      <td>0.983508</td>\n",
       "      <td>1.073539</td>\n",
       "      <td>0.775425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>S</td>\n",
       "      <td>0.916496</td>\n",
       "      <td>0</td>\n",
       "      <td>1.498919</td>\n",
       "      <td>0.796677</td>\n",
       "      <td>0.792019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>T</td>\n",
       "      <td>0.823655</td>\n",
       "      <td>0</td>\n",
       "      <td>0.853654</td>\n",
       "      <td>0.929890</td>\n",
       "      <td>0.870773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>G</td>\n",
       "      <td>0.990309</td>\n",
       "      <td>0</td>\n",
       "      <td>0.711916</td>\n",
       "      <td>0.950565</td>\n",
       "      <td>0.781286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A</td>\n",
       "      <td>0.943640</td>\n",
       "      <td>0</td>\n",
       "      <td>0.775310</td>\n",
       "      <td>0.965264</td>\n",
       "      <td>0.830205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>M</td>\n",
       "      <td>0.896964</td>\n",
       "      <td>0</td>\n",
       "      <td>0.854644</td>\n",
       "      <td>0.987284</td>\n",
       "      <td>0.923327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>C</td>\n",
       "      <td>1.097294</td>\n",
       "      <td>0</td>\n",
       "      <td>1.200829</td>\n",
       "      <td>1.040452</td>\n",
       "      <td>1.009617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>F</td>\n",
       "      <td>0.747967</td>\n",
       "      <td>0</td>\n",
       "      <td>0.919865</td>\n",
       "      <td>0.418323</td>\n",
       "      <td>0.804652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>L</td>\n",
       "      <td>0.786735</td>\n",
       "      <td>0</td>\n",
       "      <td>0.945412</td>\n",
       "      <td>0.580351</td>\n",
       "      <td>1.028494</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>V</td>\n",
       "      <td>0.915392</td>\n",
       "      <td>0</td>\n",
       "      <td>0.990329</td>\n",
       "      <td>1.041931</td>\n",
       "      <td>1.642841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>I</td>\n",
       "      <td>0.699317</td>\n",
       "      <td>0</td>\n",
       "      <td>0.942376</td>\n",
       "      <td>0.900039</td>\n",
       "      <td>1.367404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>H</td>\n",
       "      <td>1.747481</td>\n",
       "      <td>0</td>\n",
       "      <td>0.987208</td>\n",
       "      <td>0.874230</td>\n",
       "      <td>0.764390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>R</td>\n",
       "      <td>0.769954</td>\n",
       "      <td>0</td>\n",
       "      <td>0.792994</td>\n",
       "      <td>0.989830</td>\n",
       "      <td>0.746068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>K</td>\n",
       "      <td>0.579950</td>\n",
       "      <td>0</td>\n",
       "      <td>0.524937</td>\n",
       "      <td>0.903467</td>\n",
       "      <td>0.464914</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   AA        -1   Q         1         2         3\n",
       "0   D  0.627703   0  0.779207  0.795582  0.685065\n",
       "1   E  0.684350   0  0.598563  0.842955  0.543147\n",
       "2   N  0.741804   0  0.701087  0.843187  0.796826\n",
       "3   Q  1.497140  10  1.035081  1.012754  0.704695\n",
       "4   P  0.918332   0  0.567865  0.603739  0.676449\n",
       "5   Y  0.964919   0  1.106978  1.679671  0.871349\n",
       "6   W  1.093472   0  0.983508  1.073539  0.775425\n",
       "7   S  0.916496   0  1.498919  0.796677  0.792019\n",
       "8   T  0.823655   0  0.853654  0.929890  0.870773\n",
       "9   G  0.990309   0  0.711916  0.950565  0.781286\n",
       "10  A  0.943640   0  0.775310  0.965264  0.830205\n",
       "11  M  0.896964   0  0.854644  0.987284  0.923327\n",
       "12  C  1.097294   0  1.200829  1.040452  1.009617\n",
       "13  F  0.747967   0  0.919865  0.418323  0.804652\n",
       "14  L  0.786735   0  0.945412  0.580351  1.028494\n",
       "15  V  0.915392   0  0.990329  1.041931  1.642841\n",
       "16  I  0.699317   0  0.942376  0.900039  1.367404\n",
       "17  H  1.747481   0  0.987208  0.874230  0.764390\n",
       "18  R  0.769954   0  0.792994  0.989830  0.746068\n",
       "19  K  0.579950   0  0.524937  0.903467  0.464914"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score_table = pd.read_csv(score_table_file)\n",
    "score_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's take a look at the scoring matrix and confirm that it is correct, as well as help us choose a minimum_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "According to the score table, the min/max possible scores are:\n",
      "Max: 16.568911359\tHQSYV\n",
      "Min: 1.9881229449999998\tKDKFK\n"
     ]
    }
   ],
   "source": [
    "best_peptide = \"\"\n",
    "worst_peptide = \"\"\n",
    "\n",
    "for col in score_table.columns[1:]:\n",
    "    best_peptide += score_table.iloc[np.argmax(score_table[col]), 0]\n",
    "    worst_peptide += score_table.iloc[np.argmin(score_table[col]), 0]\n",
    "\n",
    "    \n",
    "max_score = score_table.iloc[:,1:].max().sum()\n",
    "min_score = score_table.iloc[:,1:].min().sum()\n",
    "\n",
    "print(\"According to the score table, the min/max possible scores are:\")\n",
    "print(f\"Max: {max_score}\\t{best_peptide}\")\n",
    "print(f\"Min: {min_score}\\t{worst_peptide}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "score_table_list = score_table.sort_values(\"AA\").values.tolist()\n",
    "n = len(score_table.columns)-1 #n is the number of amino acids in the peptide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "pep_num_list = [(\"A\", 0), (\"C\", 1), (\"D\", 2), (\"E\", 3), (\"F\", 4), (\"G\", 5), (\"H\", 6), \\\n",
    "          (\"I\", 7), (\"K\", 8), (\"L\", 9), (\"M\", 10), (\"N\", 11), (\"P\", 12), (\"Q\", 13),\\\n",
    "          (\"R\", 14), (\"S\", 15), (\"T\", 16), (\"V\", 17), (\"W\", 18), (\"Y\", 19)]\n",
    "pep_num_dict=dict(pep_num_list) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# output pep is removed at 2nd Q\n",
    "def num2pep(num, n=6):\n",
    "    \"\"\"Generates peptides of length n from numbers. Default: 6 \n",
    "    0 -> AAAAAA, 1 -> AAAAAC ... (20**6)-1 -> YYYYYY\n",
    "    \"\"\"\n",
    "    \n",
    "    base_20 = np.base_repr(num, 20)   \n",
    "    out = \"\"\n",
    "    for char in str(base_20).zfill(n):\n",
    "        base_10 = int(char, 20)        \n",
    "        out += pep_num_list[base_10][0] \n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pep2num(pep, n=6):\n",
    "    \"\"\"Converts a peptide sequence to a unique number, which allows a lookup in an array\"\"\"\n",
    "    num = 0\n",
    "    for i in range(n):\n",
    "        if(pep[i] not in pep_num_dict):\n",
    "            raise Exception(\"{} is not in pep_num_dict\".format(pep[i]))\n",
    "        num += pep_num_dict[pep[i]] * 20 **((n-1)-i) #10進法に変換 \n",
    "    return num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pep2score(pep):\n",
    "    \"\"\"Calculates the score for a given peptide.\"\"\"\n",
    "    score = 0\n",
    "    for i, item in enumerate(pep):\n",
    "        num = pep_num_dict[item]\n",
    "        score += score_table_list[num][i+1]\n",
    "    return score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#let's generate an empty array which will hold our peptide scores. For now, all scores are set to -1.\n",
    "#if we encouter a peptide for the first time in a protein, we'll calculate the score and store it in this list\n",
    "#to re-use it later. The position of a peptide in the list is calculated via pep2num()\n",
    "score_ref = [-1]*(20**n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "out = []\n",
    "\n",
    "for record in SeqIO.parse(fasta_in, 'fasta'):\n",
    "    id_part = record.id\n",
    "    desc_part = record.description\n",
    "    seq = str(record.seq)\n",
    "    \n",
    "    for i in range (len(seq)-n+1):\n",
    "        pep = seq[i:i+n]\n",
    "        \n",
    "        try:    \n",
    "            num = pep2num(pep, n)\n",
    "            score = score_ref[num]\n",
    "            if score == -1:\n",
    "                #the score has not been calculated yet\n",
    "                score = pep2score(pep)\n",
    "                score_ref[num] = score #let's store the score, to allow a quick lookup if we encouter the peptide again\n",
    "            \n",
    "        except Exception as e:\n",
    "            score = 0\n",
    "        if score > minimum_score:\n",
    "\n",
    "            seq_t = [id_part, desc_part, i+1, score, pep, seq]\n",
    "            out.append(seq_t) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's write the peptide result to a csv file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(out)>0:\n",
    "    out = pd.DataFrame(out)\n",
    "    out.columns = [\"id\",\"desc\",\"posi\",\"score\",\"target\",\"sequence\"]\n",
    "    out = out.set_index('id')\n",
    "    out = out.sort_values(by='score', ascending=False)\n",
    "    out.to_csv(csv_name)\n",
    "else:\n",
    "    print(\"No output file written: 0 hits\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>Let's also write a more compact output file (one row per protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(out)>0:\n",
    "    #re-assure that the output is sorted by peptide scores\n",
    "    out = out.sort_values(by='score', ascending=False)\n",
    "\n",
    "    out_list = []\n",
    "    out_list_columns = [\"id\", 'total score',\"highest individual peptide score\", 'peptide matches', 'unique peptide matches', 'sequence', 'peptides (peptide,score,count)']\n",
    "\n",
    "    #group by each protein\n",
    "    for protein_name, df in out.groupby(out.index):\n",
    "\n",
    "        protein_score = df['score'].sum() #total peptide score for this protein\n",
    "        peptide_matches = df['target'].count() #how many peptides matched\n",
    "        unique_peptide_matches = df['target'].drop_duplicates().count() #how many peptides matches (no duplicates)\n",
    "\n",
    "        highest_scoring_peptide = df['target'].iloc[0]\n",
    "        highest_peptide_score = df['score'].iloc[0]\n",
    "\n",
    "        #Let's write the protein sequence as lower case, \n",
    "        #except for positions where a peptide matches\n",
    "        seq = df['sequence'][0].lower()\n",
    "        for target in df['target']:\n",
    "            seq = re.sub(target, target, seq, flags=re.IGNORECASE)\n",
    "\n",
    "\n",
    "        peptide_dict = {}\n",
    "        peptide_counts = df.groupby('target').count().iloc[:,0].to_dict()\n",
    "\n",
    "        for peptide, score in zip(df['target'], df['score']):\n",
    "            peptide_dict[peptide] = {\"score\":score, 'count':peptide_counts[peptide]}\n",
    "\n",
    "        peptide_string = \"\"\n",
    "        for peptide, value in peptide_dict.items():\n",
    "            peptide_string+= f\"{peptide},{round(value['score'],3)},{value['count']}|\"\n",
    "        peptide_string = peptide_string[:-1]\n",
    "\n",
    "\n",
    "        out_list.append([protein_name, protein_score, highest_peptide_score, peptide_matches, unique_peptide_matches, seq, peptide_string])\n",
    "\n",
    "\n",
    "    dense_out = pd.DataFrame(out_list)\n",
    "    dense_out.columns = out_list_columns\n",
    "    dense_out = dense_out.sort_values(by='total score', ascending=False)\n",
    "    dense_out.to_csv(csv_name.replace(\".csv\", '.dense.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
